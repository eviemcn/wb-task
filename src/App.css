@font-face {
  font-family: "open-sans";
  font-style: normal;
  font-weight: normal;
  src: local("Open Sans Regular"), local("OpenSans-Regular"),
    url("./OpenSans-Regular.woff2") format("woff2");
}

body {
  background-color: #fff;
  display: grid;
  justify-content: center;
  font-family: "open-sans";
}

.loader {
  border: 10px solid #f3f3f3; /* Light grey */
  border-top: 10px solid #3b5eed; /* Blue */
  border-radius: 50%;
  width: 5vmin;
  height: 5vmin;
  animation: spin 1.5s linear infinite;
}

.loading {
  margin-top: 20vmin;
  text-align: center;
}

@keyframes spin {
  0% { transform: rotate(0deg); }
  100% { transform: rotate(360deg); }
}

.product-list {
  display: grid;
}

.product-list .search {
  margin: 20px 0;
  display: flex;
  align-items: center;
  width: 100%;
}

.product-list .search label {
  margin-right: 30px;
}

.product-list .search input{
  align-content: center;
  flex-grow: 1;
  border: solid 3px #d1d1d1;
  box-sizing: border-box;
  padding: 3px 8px;
  border-radius: 4px;
  line-height: 24px;
  outline: none;
}

.product-list .search input:focus{
  border: solid 3px #3b5eed;
  box-shadow: 1px 1px 8px 1px #3b5eed;
}

.product-list table {
  border-collapse: collapse;
}

.product-list thead , .product-list tfoot{
  color: #ffffff;
}

.product-list tbody {
  color: #000000;
}

.product-list td {
  /* border: 1px solid #626a80; */
  padding: 5px 15px;
}

.product-list th{
  padding: 15px;
}

.product-list th:first-child, .product-list tfoot td:first-child{
  width: 300px;
  text-align: left;
  background-color: #3b5eed;
  border-radius: 10px 0 0 10px;
}

.product-list th:last-child, .product-list tfoot td:last-child{
  width: 100px;
  text-align: right;
  background-color: #3b5eed;
  border-radius: 0 10px 10px 0;
}

.product-list tr:nth-child(2n) td:first-child{
  border-radius: 10px 0 0 10px;
}

.product-list tr:nth-child(2n) td:last-child{
  border-radius: 0 10px 10px 0;
}

.product-list tr:nth-child(2n) {
  background-color: #b7c3e4;
}

.product-list tr:nth-child(2n + 1) {
  background-color: #ffffff;
}

.product-list td:last-child {
  text-align: right;
}

