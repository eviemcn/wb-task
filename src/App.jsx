import React, { useState, useEffect } from "react";
import { getBranchData } from "./services/get-products-service.jsx";
import { ProductTable } from "./components/product-table/product-table";
import { SearchItems } from "./components/search-items/search-items";

function App() {
  const [loading, setLoading] = useState(true);
  const [branchData, setBranchData] = useState(null);
  const [searchString, setSearchString] = useState("");

  useEffect(() => {
    getBranchData().then((data) => {
      setBranchData(data);
      //Show loading spinner
      setTimeout(() => {
        setLoading(false);
      }, 500);
    });
  }, []);

  return (
    <div className="page-container">
      {loading ? (
        <div className="loading">
          <div className="loader"></div>
          <p>Loading...</p>
        </div>
      ) : (
        <div className="product-list">
          <SearchItems
            setFilter={setSearchString}
            searchString={searchString}
          />
          <ProductTable products={branchData} searchString={searchString} />
        </div>
      )}
    </div>
  );
}

export default App;
