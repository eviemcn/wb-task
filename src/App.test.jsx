import branch1 from "../public/api/branch1.json";
import branch2 from "../public/api/branch2.json";
import branch3 from "../public/api/branch3.json";
import App from "./App";
import {
  fireEvent,
  render,
  screen,
  waitFor,
  within,
} from "@testing-library/react";
import { expect, it } from "vitest";

const responses = {
  "api/branch1.json": branch1,
  "api/branch2.json": branch2,
  "api/branch3.json": branch3,
  "/api/branch1.json": branch1,
  "/api/branch2.json": branch2,
  "/api/branch3.json": branch3,
};

global.fetch = (endpoint) =>
  Promise.resolve({
    json: () => Promise.resolve(responses[endpoint]),
  });

it("renders without crashing", () => {
  render(<App />);
});

it("renders loading text initially", async () => {
  render(<App />);
  expect(screen.getByText("Loading...")).toBeInTheDocument();
});

it("renders a table after data load", async () => {
  render(<App />);
  expect(screen.getByText("Loading...")).toBeInTheDocument();
  await waitFor(() => expect(screen.getByRole("table")).toBeInTheDocument());
});

it("renders rows with product name as key", async () => {
  render(<App />);
  expect(screen.getByText("Loading...")).toBeInTheDocument();
  await waitFor(() => expect(screen.getByRole("table")).toBeInTheDocument());

  expect(screen.getAllByRole("row").at(57)).toHaveTextContent("Hominy");
  expect(screen.getAllByRole("row").at(74)).toHaveTextContent("Lychee");
});

it("renders table that is sorted ascending", async () => {
  render(<App />);
  await waitFor(() => expect(screen.getByText("Loading...")).toBeInTheDocument());
  await waitFor(() => expect(screen.getByRole("table")).toBeInTheDocument());

  expect(screen.getByRole("table")).toMatchSnapshot();
});

it("calculates total revenue of all branches", async () => {
  render(<App />);
  expect(screen.getByText("Loading...")).toBeInTheDocument();
  await waitFor(() => expect(screen.getByRole("table")).toBeInTheDocument());

  expect(
    within(screen.getAllByRole("row").at(-1)).getAllByRole("cell").at(-1)
  ).toHaveTextContent("2,102,619.44");
});

it("filters the displayed products", async () => {
  render(<App />);
  expect(screen.getByText("Loading...")).toBeInTheDocument();
  await waitFor(() => expect(screen.getByRole("table")).toBeInTheDocument());

  fireEvent.change(screen.getByLabelText("Search Products"), {
    target: { value: "pear" },
  });
  expect(
    within(screen.getAllByRole("row").at(-1)).getAllByRole("cell").at(-1)
  ).toHaveTextContent("60,681.02");
});
