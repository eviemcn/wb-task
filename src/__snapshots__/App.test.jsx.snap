// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders table that is sorted ascending 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Product
      </th>
      <th>
        Revenue
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        Alfalfa Sprouts
      </td>
      <td>
        12,486.74
      </td>
    </tr>
    <tr>
      <td>
        Apple
      </td>
      <td>
        14,832.72
      </td>
    </tr>
    <tr>
      <td>
        Apricot
      </td>
      <td>
        20,647.26
      </td>
    </tr>
    <tr>
      <td>
        Artichoke
      </td>
      <td>
        4,849.95
      </td>
    </tr>
    <tr>
      <td>
        Asian Pear
      </td>
      <td>
        11,601.45
      </td>
    </tr>
    <tr>
      <td>
        Asparagus
      </td>
      <td>
        20,105.02
      </td>
    </tr>
    <tr>
      <td>
        Atemoya
      </td>
      <td>
        11,160.60
      </td>
    </tr>
    <tr>
      <td>
        Avocado
      </td>
      <td>
        8,842.50
      </td>
    </tr>
    <tr>
      <td>
        Bamboo Shoots
      </td>
      <td>
        11,289.60
      </td>
    </tr>
    <tr>
      <td>
        Banana
      </td>
      <td>
        10,872.52
      </td>
    </tr>
    <tr>
      <td>
        Bean Sprouts
      </td>
      <td>
        16,626.56
      </td>
    </tr>
    <tr>
      <td>
        Beans
      </td>
      <td>
        21,774.33
      </td>
    </tr>
    <tr>
      <td>
        Beets
      </td>
      <td>
        25,521.00
      </td>
    </tr>
    <tr>
      <td>
        Belgian Endive
      </td>
      <td>
        14,682.80
      </td>
    </tr>
    <tr>
      <td>
        Bell Peppers
      </td>
      <td>
        24,284.16
      </td>
    </tr>
    <tr>
      <td>
        Bitter Melon
      </td>
      <td>
        17,095.69
      </td>
    </tr>
    <tr>
      <td>
        Blackberries
      </td>
      <td>
        28,646.27
      </td>
    </tr>
    <tr>
      <td>
        Blueberries
      </td>
      <td>
        12,556.70
      </td>
    </tr>
    <tr>
      <td>
        Bok Choy
      </td>
      <td>
        20,254.18
      </td>
    </tr>
    <tr>
      <td>
        Boniato
      </td>
      <td>
        9,098.32
      </td>
    </tr>
    <tr>
      <td>
        Boysenberries
      </td>
      <td>
        12,843.36
      </td>
    </tr>
    <tr>
      <td>
        Broccoflower
      </td>
      <td>
        33,651.20
      </td>
    </tr>
    <tr>
      <td>
        Broccoli
      </td>
      <td>
        19,440.30
      </td>
    </tr>
    <tr>
      <td>
        Brussels Sprouts
      </td>
      <td>
        3,769.68
      </td>
    </tr>
    <tr>
      <td>
        Cabbage
      </td>
      <td>
        11,514.00
      </td>
    </tr>
    <tr>
      <td>
        Cactus Pear
      </td>
      <td>
        21,227.68
      </td>
    </tr>
    <tr>
      <td>
        Cantaloupe
      </td>
      <td>
        14,853.96
      </td>
    </tr>
    <tr>
      <td>
        Carambola
      </td>
      <td>
        9,796.99
      </td>
    </tr>
    <tr>
      <td>
        Carrots
      </td>
      <td>
        29,493.04
      </td>
    </tr>
    <tr>
      <td>
        Casaba Melon
      </td>
      <td>
        6,220.17
      </td>
    </tr>
    <tr>
      <td>
        Cauliflower
      </td>
      <td>
        10,843.26
      </td>
    </tr>
    <tr>
      <td>
        Celery
      </td>
      <td>
        9,718.22
      </td>
    </tr>
    <tr>
      <td>
        Chayote
      </td>
      <td>
        20,305.76
      </td>
    </tr>
    <tr>
      <td>
        Cherimoya
      </td>
      <td>
        15,324.21
      </td>
    </tr>
    <tr>
      <td>
        Cherries
      </td>
      <td>
        7,505.74
      </td>
    </tr>
    <tr>
      <td>
        Coconuts
      </td>
      <td>
        9,075.53
      </td>
    </tr>
    <tr>
      <td>
        Collard Greens
      </td>
      <td>
        16,758.56
      </td>
    </tr>
    <tr>
      <td>
        Corn
      </td>
      <td>
        9,437.04
      </td>
    </tr>
    <tr>
      <td>
        Cranberries
      </td>
      <td>
        28,912.94
      </td>
    </tr>
    <tr>
      <td>
        Cucumber
      </td>
      <td>
        12,106.82
      </td>
    </tr>
    <tr>
      <td>
        Dates
      </td>
      <td>
        18,813.60
      </td>
    </tr>
    <tr>
      <td>
        Dried Plums
      </td>
      <td>
        10,022.30
      </td>
    </tr>
    <tr>
      <td>
        Eggplant
      </td>
      <td>
        10,173.60
      </td>
    </tr>
    <tr>
      <td>
        Endive
      </td>
      <td>
        15,671.01
      </td>
    </tr>
    <tr>
      <td>
        Escarole
      </td>
      <td>
        8,443.37
      </td>
    </tr>
    <tr>
      <td>
        Feijoa
      </td>
      <td>
        12,015.94
      </td>
    </tr>
    <tr>
      <td>
        Fennel
      </td>
      <td>
        12,870.70
      </td>
    </tr>
    <tr>
      <td>
        Figs
      </td>
      <td>
        14,169.36
      </td>
    </tr>
    <tr>
      <td>
        Garlic
      </td>
      <td>
        16,872.96
      </td>
    </tr>
    <tr>
      <td>
        Gooseberries
      </td>
      <td>
        2,980.52
      </td>
    </tr>
    <tr>
      <td>
        Grapefruit
      </td>
      <td>
        20,697.72
      </td>
    </tr>
    <tr>
      <td>
        Grapes
      </td>
      <td>
        26,356.69
      </td>
    </tr>
    <tr>
      <td>
        Green Beans
      </td>
      <td>
        11,531.14
      </td>
    </tr>
    <tr>
      <td>
        Green Onions
      </td>
      <td>
        15,536.90
      </td>
    </tr>
    <tr>
      <td>
        Greens
      </td>
      <td>
        11,121.90
      </td>
    </tr>
    <tr>
      <td>
        Guava
      </td>
      <td>
        28,212.45
      </td>
    </tr>
    <tr>
      <td>
        Hominy
      </td>
      <td>
        9,083.96
      </td>
    </tr>
    <tr>
      <td>
        Honeydew Melon
      </td>
      <td>
        10,559.43
      </td>
    </tr>
    <tr>
      <td>
        Horned Melon
      </td>
      <td>
        6,569.64
      </td>
    </tr>
    <tr>
      <td>
        Iceberg Lettuce
      </td>
      <td>
        21,069.78
      </td>
    </tr>
    <tr>
      <td>
        Jerusalem Artichoke
      </td>
      <td>
        28,203.00
      </td>
    </tr>
    <tr>
      <td>
        Jicama
      </td>
      <td>
        13,444.20
      </td>
    </tr>
    <tr>
      <td>
        Kale
      </td>
      <td>
        8,933.76
      </td>
    </tr>
    <tr>
      <td>
        Kiwifruit
      </td>
      <td>
        33,122.60
      </td>
    </tr>
    <tr>
      <td>
        Kohlrabi
      </td>
      <td>
        21,898.89
      </td>
    </tr>
    <tr>
      <td>
        Kumquat
      </td>
      <td>
        13,267.92
      </td>
    </tr>
    <tr>
      <td>
        Leeks
      </td>
      <td>
        7,269.34
      </td>
    </tr>
    <tr>
      <td>
        Lemons
      </td>
      <td>
        17,755.07
      </td>
    </tr>
    <tr>
      <td>
        Lettuce
      </td>
      <td>
        10,037.56
      </td>
    </tr>
    <tr>
      <td>
        Lima Beans
      </td>
      <td>
        3,670.68
      </td>
    </tr>
    <tr>
      <td>
        Limes
      </td>
      <td>
        15,708.00
      </td>
    </tr>
    <tr>
      <td>
        Longan
      </td>
      <td>
        18,495.20
      </td>
    </tr>
    <tr>
      <td>
        Loquat
      </td>
      <td>
        5,705.68
      </td>
    </tr>
    <tr>
      <td>
        Lychee
      </td>
      <td>
        20,981.66
      </td>
    </tr>
    <tr>
      <td>
        Madarins
      </td>
      <td>
        4,747.44
      </td>
    </tr>
    <tr>
      <td>
        Malanga
      </td>
      <td>
        18,386.10
      </td>
    </tr>
    <tr>
      <td>
        Mandarin Oranges
      </td>
      <td>
        12,380.40
      </td>
    </tr>
    <tr>
      <td>
        Mangos
      </td>
      <td>
        10,977.28
      </td>
    </tr>
    <tr>
      <td>
        Mulberries
      </td>
      <td>
        9,189.96
      </td>
    </tr>
    <tr>
      <td>
        Mushrooms
      </td>
      <td>
        12,557.84
      </td>
    </tr>
    <tr>
      <td>
        Napa
      </td>
      <td>
        15,087.56
      </td>
    </tr>
    <tr>
      <td>
        Nectarines
      </td>
      <td>
        31,706.88
      </td>
    </tr>
    <tr>
      <td>
        Okra
      </td>
      <td>
        20,784.21
      </td>
    </tr>
    <tr>
      <td>
        Onion
      </td>
      <td>
        15,888.04
      </td>
    </tr>
    <tr>
      <td>
        Oranges
      </td>
      <td>
        10,575.18
      </td>
    </tr>
    <tr>
      <td>
        Papayas
      </td>
      <td>
        19,369.74
      </td>
    </tr>
    <tr>
      <td>
        Parsnip
      </td>
      <td>
        17,594.65
      </td>
    </tr>
    <tr>
      <td>
        Passion Fruit
      </td>
      <td>
        21,589.92
      </td>
    </tr>
    <tr>
      <td>
        Peaches
      </td>
      <td>
        18,759.18
      </td>
    </tr>
    <tr>
      <td>
        Pears
      </td>
      <td>
        10,329.30
      </td>
    </tr>
    <tr>
      <td>
        Peas
      </td>
      <td>
        17,965.60
      </td>
    </tr>
    <tr>
      <td>
        Peppers
      </td>
      <td>
        8,838.01
      </td>
    </tr>
    <tr>
      <td>
        Persimmons
      </td>
      <td>
        24,514.18
      </td>
    </tr>
    <tr>
      <td>
        Pineapple
      </td>
      <td>
        23,462.40
      </td>
    </tr>
    <tr>
      <td>
        Plantains
      </td>
      <td>
        9,978.90
      </td>
    </tr>
    <tr>
      <td>
        Plums
      </td>
      <td>
        13,314.60
      </td>
    </tr>
    <tr>
      <td>
        Pomegranate
      </td>
      <td>
        13,551.09
      </td>
    </tr>
    <tr>
      <td>
        Potatoes
      </td>
      <td>
        10,812.54
      </td>
    </tr>
    <tr>
      <td>
        Prickly Pear
      </td>
      <td>
        17,522.59
      </td>
    </tr>
    <tr>
      <td>
        Prunes
      </td>
      <td>
        13,644.54
      </td>
    </tr>
    <tr>
      <td>
        Pummelo
      </td>
      <td>
        20,429.40
      </td>
    </tr>
    <tr>
      <td>
        Pumpkin
      </td>
      <td>
        10,547.88
      </td>
    </tr>
    <tr>
      <td>
        Quince
      </td>
      <td>
        15,250.34
      </td>
    </tr>
    <tr>
      <td>
        Radicchio
      </td>
      <td>
        20,453.27
      </td>
    </tr>
    <tr>
      <td>
        Radishes
      </td>
      <td>
        30,051.48
      </td>
    </tr>
    <tr>
      <td>
        Raisins
      </td>
      <td>
        7,411.05
      </td>
    </tr>
    <tr>
      <td>
        Raspberries
      </td>
      <td>
        4,318.82
      </td>
    </tr>
    <tr>
      <td>
        Red Cabbage
      </td>
      <td>
        12,005.60
      </td>
    </tr>
    <tr>
      <td>
        Rhubarb
      </td>
      <td>
        14,905.80
      </td>
    </tr>
    <tr>
      <td>
        Romaine Lettuce
      </td>
      <td>
        32,786.28
      </td>
    </tr>
    <tr>
      <td>
        Rutabaga
      </td>
      <td>
        18,551.12
      </td>
    </tr>
    <tr>
      <td>
        Shallots
      </td>
      <td>
        6,517.17
      </td>
    </tr>
    <tr>
      <td>
        Snow Peas
      </td>
      <td>
        11,989.19
      </td>
    </tr>
    <tr>
      <td>
        Spinach
      </td>
      <td>
        19,822.32
      </td>
    </tr>
    <tr>
      <td>
        Sprouts
      </td>
      <td>
        27,061.80
      </td>
    </tr>
    <tr>
      <td>
        Squash
      </td>
      <td>
        15,080.00
      </td>
    </tr>
    <tr>
      <td>
        Strawberries
      </td>
      <td>
        11,681.46
      </td>
    </tr>
    <tr>
      <td>
        String Beans
      </td>
      <td>
        22,731.62
      </td>
    </tr>
    <tr>
      <td>
        Sweet Potato
      </td>
      <td>
        31,680.00
      </td>
    </tr>
    <tr>
      <td>
        Tangelo
      </td>
      <td>
        19,831.35
      </td>
    </tr>
    <tr>
      <td>
        Tangerines
      </td>
      <td>
        29,450.24
      </td>
    </tr>
    <tr>
      <td>
        Tomatillo
      </td>
      <td>
        9,560.46
      </td>
    </tr>
    <tr>
      <td>
        Tomato
      </td>
      <td>
        14,517.90
      </td>
    </tr>
    <tr>
      <td>
        Turnip
      </td>
      <td>
        15,564.12
      </td>
    </tr>
    <tr>
      <td>
        Ugli Fruit
      </td>
      <td>
        10,838.10
      </td>
    </tr>
    <tr>
      <td>
        Water Chestnuts
      </td>
      <td>
        7,857.66
      </td>
    </tr>
    <tr>
      <td>
        Watercress
      </td>
      <td>
        13,642.14
      </td>
    </tr>
    <tr>
      <td>
        Watermelon
      </td>
      <td>
        19,219.20
      </td>
    </tr>
    <tr>
      <td>
        Waxed Beans
      </td>
      <td>
        9,119.88
      </td>
    </tr>
    <tr>
      <td>
        Yams
      </td>
      <td>
        13,719.66
      </td>
    </tr>
    <tr>
      <td>
        Yellow Squash
      </td>
      <td>
        23,634.52
      </td>
    </tr>
    <tr>
      <td>
        Yuca/Cassava
      </td>
      <td>
        44,490.84
      </td>
    </tr>
    <tr>
      <td>
        Zucchini Squash
      </td>
      <td>
        13,077.28
      </td>
    </tr>
  </tbody>
  <tfoot>
    <tr>
      <td>
        Total
      </td>
      <td>
        2,102,619.44
      </td>
    </tr>
  </tfoot>
</table>
`;
